File: problems/0028-find-the-index-of-the-first-occurrence-in-a-string/solution_test.go
package findtheindexofthefirstoccurrenceinastring

import (
	"reflect"
	"testing"
)

func TestFindTheIndexOfTheFirstOccurrenceInAString(t *testing.T) {
	tests := []struct {
		name     string
		haystack string
		needle   string
		want     int
	}{
		{"empty", "", "sad", -1},
		{"single-negative", "s", "as", -1},
		{"single-positive", "s", "s", 0},
		{"case1", "sadbutsad", "sad", 0},
		{"case2", "leetcode", "leeto", -1},
		{"case3", "hello", "ll", 2},
		{"case4", "mississippi", "issip", 4},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := FindTheIndexOfTheFirstOccurrenceInAString(tt.haystack, tt.needle)
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("RemoveDuplicatesFromSortedArray(%s, %s) = %d, want %d", tt.haystack, tt.needle, got, tt.want)
			}
		})
	}
}

File: problems/0028-find-the-index-of-the-first-occurrence-in-a-string/solution.test.ts
import { describe, expect, it } from "vitest";
import { findTheIndexOfTheFirstOccurrenceInAString } from "./solution";
// import { findTheIndexOfTheFirstOccurrenceInAString } from "./solution";

describe("findTheIndexOfTheFirstOccurrenceInAString", () => {
  it("empty", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("", "sad")).toEqual(-1);
  });

  it("single-negative", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("s", "as")).toEqual(-1);
  });

  it("single-positive", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("s", "s")).toEqual(0);
  });

  it("case1", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("sadbutsad", "sad")).toEqual(0);
  });

  it("case2", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("leetcode", "leeto")).toEqual(-1);
  });

  it("case3", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("hello", "ll")).toEqual(2);
  });

  it("case4", () => {
    expect(findTheIndexOfTheFirstOccurrenceInAString("mississippi", "issip")).toEqual(4);
  });
});
